core/arch/arch.c

```diff
@@ -3633,8 +3633,35 @@ priv_mcontext_to_dr_mcontext(dr_mcontext_t *dst, priv_mcontext_t *src)
                     return false;
                 memcpy(&dst->opmask, &src->opmask, sizeof(dst->opmask));
             }
+#elif defined(AARCHXX)
+#    ifdef X64
+            /* The first and so far (August 2023) only DR_MC_MULTIMEDIA related
+             * change to AARCH64's machine context was addition of the Scalable
+             * Vector Extension (SVE) vector registers (DR_REG_Z0->DR_REG_Z31),
+             * predicate registers (DR_REG_P0->DR_REG_P15) and first fault
+             * register (DR_REG_FFR). Further SVE2 support did not change
+             * these. To avoid breaking backward compatibility with pre-SVE
+             * clients, we only copy the first 128 bits of each vector, i.e.
+             * the v8 NEON SIMD registers DR_REG_Q0->DR_REG_Q31.
+             */
+            if (sizeof(dst->simd) < sizeof(src->simd)) {
```

Comment:
As the `ASSERT_NOT_IMPLEMENTED(false)` is not being hit by any current Arm tests on SVE and pre-SVE hardware, can we separate this binary compatibility issue into another PR? This one is big enough as it is and IMHO it would be better to split it out as a separate piece of work, as with the signal context changes.

We could do with this patch in the master so we can unblock scatter/gather work (https://github.com/DynamoRIO/dynamorio/issues/5844) as well as flushing out any SVE bugs or pre-SVE regressions in a timely manner. 

If we require users to rebuild their clients/tools for SVE compatibility after release 10.0, we don't have to do any binary compatibility changes for pre-SVE clients/tools.

Reply:
> As the ASSERT_NOT_IMPLEMENTED(false) is not being hit by any current Arm tests on SVE and pre-SVE hardware

But it *will* be hit by anyone using DR_MC_ALL who hasn't recompiled because the size won't match so it won't go into the single copy case and will then go into the DR_MC_MULTIMEDIA code.  Let's add a check up top for the user size not matching and return false in that case for AARCH64.